common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java

```diff
@@ -203,16 +215,19 @@ private synchronized boolean shouldRetry(Throwable e) {
     boolean isIOException = e instanceof IOException
       || e.getCause() instanceof IOException;
     boolean isSaslTimeout = enableSaslRetries && e instanceof SaslTimeoutException;
-    if (!isSaslTimeout && saslTimeoutSeen) {
-      retryCount = 0;
-      saslTimeoutSeen = false;
+    // If this is a non SASL request failure, reduce earlier SASL failures from retryCount
+    // since some subsequent SASL attempt was successful
+    if (!isSaslTimeout && saslRetryCount > 0) {
+      retryCount -= saslRetryCount;
+      if (retryCount < 0) {
+        throw new RuntimeException(
+          "retryCount is negative: " + retryCount + "(sasl retry count: " + saslRetryCount + ").");
+      }
```

Comment:
Use `Preconditions` instead - `Preconditions.checkState(retryCount >= saslRetryCount)` before the subtraction.